### ISSUE TITLE
Issue with Modularity

### ISSUE THREAD

**Issue by Jonty800** [gephi/gephi-toolkit#20]
I'm reading a .gexf and loading 126 nodes with 378 edges. I'm then trying to perform Modularity:
 `Modularity modularity = new Modularity();
        modularity.setRandom(true);
        modularity.setUseWeight(true);
        modularity.setResolution(0.8);
        modularity.setProgressTicket(ticket);
        modularity.execute(graphModel);`

However, the algorithm hangs. Inside ProgressTicket, only start() is called and finish() is never called. It hangs for about 2 minutes and then stops "Process finished with exit code 0". - No errors.

What is going on here?

**Comment by eduramiba** [comment]
Hi, this is already reported at https://github.com/gephi/gephi/issues/1630 and is in the list of priority fixes.

**Comment by Jonty800** [comment]
Oh I see. It works perfectly in the desktop application, just not in the toolkit.

Do you have any suggestions on how I could maybe get it to work? If I can automate the modularity it would speed my work up a lot :)

**Comment by eduramiba** [comment]
We still need to find the bug, though it should not work for you in desktop app either. You can try to avoid parallel edges, since maybe this is other bug already fixed in desktop but not released in toolkit.

**Comment by Jonty800** [comment]
I use a java program to build my .gexf and use this file for both the toolkit and the app. I'll run some more tests tomorrow, and ensure there are no parallel edges. Thanks for the help!

**Comment by Jonty800** [comment]
I've opened my .gexf in the Gephi app (0.9.1) and I can see that it is merging the parallel weights. I then save this file from Gephi app to a new .gexf.

When I use this new .gexf, the hanging issue no longer occurs, however, only start() is called and never finish(). "Process finished with exit code 0" almost immediately.

I can't identify why the Gephi app works, and the toolkit does not.

I hope this helps - good luck with the bug :)

**Comment by eduramiba** [comment]
That's strange, can you show your java code? Or compare with this example https://github.com/gephi/gephi-toolkit-demos/blob/master/src/main/java/org/gephi/toolkit/demos/HeadlessSimple.java

**Comment by Jonty800** [comment]
Sure: https://gist.github.com/Jonty800/d284ff55fb55cd42bb49622b90052ba9

**Comment by eduramiba** [comment]
Mmm can you try to pass the graphModel instead of graph like in https://github.com/gephi/gephi-toolkit-demos/blob/master/src/main/java/org/gephi/toolkit/demos/PartitionGraph.java ?

**Comment by Jonty800** [comment]
Yup, I tried with both graph and graphModel since you can use either. The same happens with both.

**Comment by eduramiba** [comment]
I see, you can also try to use the latest 0.9.2 snapshot build:

https://github.com/gephi/gephi-toolkit#nightly-builds-092-snapshot

**Comment by Jonty800** [comment]
I tried this a couple of days ago with gephi-toolkit-0.9.2-20170504.212034-186-all - same thing

**Comment by eduramiba** [comment]
Ok, will you mail me your test file for reproducing it please? Or attach a zip here if you can

**Comment by Jonty800** [comment]
[gephi files.zip](https://github.com/gephi/gephi-toolkit/files/981628/gephi.files.zip)

Sure -> here is the .gexf produced by my java code, and the same file but opened and then saved by gephi

**Comment by eduramiba** [comment]
Thanks! I will check it later

**Comment by eduramiba** [comment]
This seems to work with latest toolkit 0.9.2-SNAPSHOT. Can you also test it please?

**Comment by Jonty800** [comment]
Sorry for the late reply, it's been a while since I've needed to use Gephi.

I've been testing the 0.9.3-SNAPSHOT of the toolkit and I seem to get the same problem (only start() method is called, and then it hangs).

My gexf is created using java code only. There are duplicates (A vs B, B vs A), as you can see, for edges:

`<?xml version='1.0' encoding='UTF-8'?>
<gexf xmlns="http://www.gexf.net/1.3" version="1.3" xmlns:viz="http://www.gexf.net/1.3/viz" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.gexf.net/1.3 http://www.gexf.net/1.3/gexf.xsd">
  <meta lastmodifieddate="2018-04-08">
    <creator>Gephi 0.9</creator>
    <description></description>
  </meta>
  <graph defaultedgetype="directed" timeformat="double" timerepresentation="interval" mode="dynamic">
    <attributes class="edge" mode="dynamic">
      <attribute id="weight" title="Weight" type="double"/>
    </attributes>
    <nodes>
      <node id="6944cc860b20cfc8" label="6944cc860b20cfc8">
        <viz:size value="0.0"/>
      </node>
      <node id="933e92ddabdfdc7d" label="933e92ddabdfdc7d">
        <viz:size value="0.0"/>
      </node>
    </nodes>
    <edges>
      <edge id="0" source="6944cc860b20cfc8" target="933e92ddabdfdc7d">
        <attvalues>
          <attvalue for="weight" value="0.2" start="0.0" end="1.0"/>
          <attvalue for="weight" value="0.2" start="1.0" end="2.0"/>
          <attvalue for="weight" value="0.09090909090909091" start="2.0" end="3.0"/>
          <attvalue for="weight" value="0.2" start="3.0" end="4.0"/>
          <attvalue for="weight" value="0.2" start="4.0" end="5.0"/>
        </attvalues>
      </edge>
      <edge id="1" source="933e92ddabdfdc7d" target="6944cc860b20cfc8">
        <attvalues>
          <attvalue for="weight" value="0.2" start="0.0" end="1.0"/>
          <attvalue for="weight" value="0.2" start="1.0" end="2.0"/>
          <attvalue for="weight" value="0.09090909090909091" start="2.0" end="3.0"/>
          <attvalue for="weight" value="0.2" start="3.0" end="4.0"/>
          <attvalue for="weight" value="0.2" start="4.0" end="5.0"/>
        </attvalues>
      </edge>
    </edges>
  </graph>
</gexf>`

Any thoughts?

**Comment by Jonty800** [comment]
I modified the Gephi code for Modularity.java and stopped it from being asynchronous. I did this by removing 

`implements Statistics, LongTask`

and all associated code from these packages:

`import org.gephi.statistics.spi.Statistics;`
 `import org.gephi.utils.longtask.spi.LongTask; ` 
 `import org.gephi.utils.progress.Progress;  `
` import org.gephi.utils.progress.ProgressTicket;`

 The class now works. I hope this helps you identify the bug.

https://gist.github.com/Jonty800/999425fbc1061fc2721ad4fc202b8175

Jon
